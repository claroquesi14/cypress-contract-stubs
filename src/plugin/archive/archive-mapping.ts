import { ArchiveEntry, StubEntries } from './archive-entry-model';
import { list, readFile } from 'ls-archive';
import { parse, extname } from 'path';
import { promisify } from 'util';

/**
 * Extract all stub entries from an archive file
 * @param archivePath A valid path for the file.
 * @returns A Promise for the completion of all stub entries.
 */
export async function archiveMapping(archivePath: string): Promise<StubEntries> {
  const entries = (await promisify(list)(archivePath))
    .filter((entry: ArchiveEntry) => extname(entry.path) === '.json')
    .map(async (entry: ArchiveEntry) => {
      const stub = JSON.parse((await promisify(readFile)(archivePath, entry.path)).toString());
      stub.name = parse(entry.path).name;

      try {
        stub.response.body = JSON.parse(stub.response.body);
        // eslint-disable-next-line no-empty
      } catch (e) {}

      return stub;
    });

  return Promise.all(entries);
}
