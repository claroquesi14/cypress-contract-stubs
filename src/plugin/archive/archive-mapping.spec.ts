import { archiveMapping } from './archive-mapping';
import { expect, jest } from '@jest/globals';
import * as ls from 'ls-archive';
import * as util from 'util';

jest.mock('ls-archive');
jest.mock('util');

describe('ArchiveMapping Test', () => {
  const archivePath = '/path/archive/stub.jar';
  const archiveList = [{ path: 'internal/location/stub.json' }];
  const archiveContent = JSON.stringify({
    id: '123456',
    request: { url: 'url/stub' },
    response: { body: JSON.stringify([{ value: true }]) }
  });

  const promisified = jest.fn();

  beforeEach(() => {
    promisified.mockResolvedValueOnce(archiveList as never).mockResolvedValueOnce(archiveContent as never);
    jest.spyOn(util, 'promisify').mockReturnValue(promisified);
  });

  it('should extract all stub entries from an archive file', () => {
    archiveMapping(archivePath);
    expect(util.promisify).toHaveBeenNthCalledWith(1, ls.list);
    expect(promisified).toHaveBeenNthCalledWith(1, archivePath);
  });

  it('should map all json stub entries', async () => {
    const mappedArchive = await archiveMapping(archivePath);
    expect(util.promisify).toHaveBeenNthCalledWith(2, ls.readFile);
    expect(promisified).toHaveBeenNthCalledWith(2, archivePath, archiveList[0].path);
    expect(mappedArchive).toMatchSnapshot();
  });

  it('should throw exception in case of archive fail', async () => {
    promisified.mockReset().mockRejectedValue('error');
    expect.assertions(1);

    try {
      await archiveMapping(archivePath);
    } catch (error) {
      expect(error).toBe('error');
    }
  });
});
