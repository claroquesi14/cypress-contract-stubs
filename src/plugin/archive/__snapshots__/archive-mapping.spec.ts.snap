// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArchiveMapping Test should map all json stub entries 1`] = `
Array [
  Object {
    "id": "123456",
    "name": "stub",
    "request": Object {
      "url": "url/stub",
    },
    "response": Object {
      "body": Array [
        Object {
          "value": true,
        },
      ],
    },
  },
]
`;
